template.yaml

```diff
@@ -1011,7 +1012,167 @@ Resources:
+          const response = require('cfn-response');
+          exports.handler = async (event, context) => {
+          context.callbackWaitsForEmptyEventLoop = true;
+            console.log("REQUEST RECEIVED:\n" + JSON.stringify(event));
+            const client = new MarketplaceCatalogClient({ region: 'us-east-1' });
+            const productId = event.ResourceProperties.ProductId; // Assuming the product ID is passed as an event parameter
+
+            try {
+                if (event.RequestType === 'Create' || event.RequestType === 'Update') {
+                    // Fetch the product details from AWS Marketplace
+                    const command = new DescribeEntityCommand({
+                        Catalog: 'AWSMarketplace',
+                        EntityId: productId,
+                        EntityType: 'Product'
+                    });
+                    const resp = await client.send(command);
+
+                    // Extract the product code from the response
+                    const productCode = resp.DetailsDocument.Description.ProductCode;
+
+                    const responseData = {
+                        ProductCode: productCode
+                    };
+
+                    await response.send(event, context, 'SUCCESS', responseData);
+                } else if (event.RequestType === 'Delete') {
+                    // No action needed for delete
+                    await response.send(event, context, 'SUCCESS', {});
+                } else {
+                    await response.send(event, context, 'FAILED', { error: 'Invalid request type' });
+                }
+            } catch (error) {
+                console.error('Error:', error);
+                await response.send(event, context, 'FAILED', { error: 'Failed to fetch product code' });
+            }
+          };
+
+  CAPILambdasExecutionRole:
+    Type: AWS::IAM::Role
+    Properties:
+      AssumeRolePolicyDocument:
+        Version: '2012-10-17'
+        Statement:
+          - Effect: Allow
+            Principal:
+              Service:
+                - lambda.amazonaws.com
+            Action:
+              - 'sts:AssumeRole'
+      ManagedPolicyArns:
+        - arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole
+        - arn:aws:iam::aws:policy/AWSMarketplaceSellerFullAccess
+
+  UpdateFulfillmentURL:
+    Type: Custom::Lambda
+    Properties:
+      ServiceToken: !GetAtt UpdateFulfillmentURLCustomResource.Arn
+      ProductId: !Ref ProductId
+      # FulfillmentUrl: 'https://cachicamo.org'
+      FulfillmentUrl: !If [
+                        CreateWeb,
+                        !Sub "https://${CloudfrontDistribution.DomainName}/redirectmarketplacetoken",
+                        !Sub "https://${ServerlessApi}.execute-api.${AWS::Region}.amazonaws.com/Prod/redirectmarketplacetoken"
+                      ]
+
+  UpdateFulfillmentURLCustomResource:
```

Comment:
Can we add a condition and parameterize this action and set the parameter default to NOT update. Someone testing this may inadvertently edit a public listing. We need to add something like "WARNING: Selecting 'Yes, update product listing' will update the product in your account automatically with the new fulfillment URL. Be careful doing this if your product is already live."